enhancements/network/ingress-node/ingress-node-firewall-with-ebpf.md

```diff
@@ -0,0 +1,480 @@
+
+### Non-Goals
+- Flood Attacks
+
+## Proposal
+
+- customers will provision `NodeEndPoint` objects where the nodeFirewall policy will be applied to, if the list is empty the policy will be applied on all interfaces.
+- customers will label nodes where ingress node policy will be deployed on.
+- customers will prepare `IngressNodeFirewall` objects with all required firewall rules and deploy, and via annotation will select the `NodeEndPoint` where the policy will be applied to, if there is no annotation rules will be attached to all interfaces.
+- IngressNodeFW controller will process `IngressNodeFirewall` objects, making sure configuration doesn't contradict with fail safe rules, update eBPF ingress table and attach eBPF kernel hook loaded as `BPF_PROG_TYPE_XDP` to the list of interfaces configured in `NodeEndPoint` object.
+- in order to fail safe and avoid cluster lockup, various pinholes will be open in the firewall to allow basic services.
+
+### Workflow Description
+
+- the following diagram shows how node ingress firewall configuration is handled
+
+![Ingress Firewall configuration processing](./ingress-nodefw-config-flow.png)
+
+- the following diagram shows eBPF how ingress firewall kernel hook will process packets and apply firewall rules
+
+![eBPF workflow](./ebpf-packet-flow.png)
+
+- the following table shows sample of fail safe rules to make sure users won't block basic cluster services by configuring invalid firewall rules
+
+| Source CIDR | Protocol | Port | Purpose |
+|-------------|----------|------|---------|
+|  0.0.0.0/0  |   TCP    |  22  |  SSH    |
+|  0.0.0.0/0  |   UDP    |  68  | DHCP    |
+|  0.0.0.0/0  |   TCP    | 6443 | Kubernetes API server access |
+|  0.0.0.0/0  |   TCP    | 2380 & 2379 | Etcd access |
+
+### Implementation Details/Notes/Constraints
+
+New downstream OLM operator (openshift/ingress-nodefw-operator) will be designed to do the following:-
+- APIs for the new objecs.
+- controller for each of the new resources to handle add/update/delete
+- verification webhook to validate the configuration.  
+- library to construct and load eBPF ingess maps tables based on the provisioned rules.
+- library for statistics collection and exporting to prometheus.
+
+New downstream repo for the kernel hook implementation (openshift/ingress-nodefw-kernel)
+
+- sample `NodeEndPoint` object
+
+```yaml
+kind: NodeEndpoint
+metadata:
+  name: worker-all-interfaces
+  labels:
+    role: k8s-worker
+spec:
+  interfaces: 
+
+---
+
+kind: NodeEndpoint
+metadata:
+  name: worker-eth0
+  labels:
+    role: k8s-worker
+spec:
+  interfaces: 
+  - "eth0"
+
+---
+```
+
+- sample `IngressNodeFirewall` object
+
+```yaml
+kind: IngressNodeFirewall
+metadata:
+  name: worker-nodes-firewall-on-eth0
+  annotations:
+    node-endpoint: worker-eth0
+spec:
+  nodeSelector: has(do-node-ingress-firewall)
+  ingress:
+  - fromCIDRS:
+       - 1.1.1.1/24
+       - 100:1::1/64
+    rules:
+    - order: 10  
```

Comment:
make this "priority"

Reply:
priority has some confusion to it is low value higher or lower priority why order is more easier to understand rules execution will be done in ascending order